File: files/000-099/005/CH-005.R
library(tidyverse)
library(readxl)

path = "files/CH-005.xlsx"
input = read_excel(path, range = "B2:D21")
test  = read_excel(path, range = "I2:L6") %>% as.matrix() 
# replace NA with 0
test[is.na(test)] = 0

result = input %>%
  mutate(value = 1) %>%
  select(-Quantity) %>%
  pivot_wider(names_from = Product, values_from = value, values_fill = list(value = 0)) %>%
  select(-c(`Invoice Num`))

prod = as.matrix(result)
prod_t = as.matrix(result) %>% t()
cooc = prod_t %*% prod

diag(cooc) = 0
rownames(cooc) = rownames(test) = colnames(test) = colnames(cooc) = colnames(result) 

identical(cooc, test)
# [1] TRUE

File: files/000-099/005/CH-005.py
import numpy as np
import pandas as pd

path = "CH-005.xlsx"
input_data = pd.read_excel(path, usecols="B:D", skiprows=1, nrows=20)
test = pd.read_excel(path, usecols="I:L", skiprows=1, nrows=4).fillna(0).astype(int).to_numpy()

prod = (
    input_data.assign(value=1)
    .drop(columns=["Quantity"])
    .pivot_table(index="Invoice Num", columns="Product", values="value", fill_value=0, aggfunc="first")
)
cooc = prod.to_numpy().T @ prod.to_numpy()
np.fill_diagonal(cooc, 0)

print(np.array_equal(cooc, test))
